vision/mlp_mnist/mlp_mnist.jl

```diff
@@ -1,174 +1,116 @@
-# * **Creates mini-batches of data:** Creates two DataLoader objects (train and test) that handle data mini-batches of size `1024 ` (as defined above). We create these two objects so that we can pass the entire data set through the loss function at once when training our model. Also, it shuffles the data points during each iteration (`shuffle=true`).
+simple_accuracy(model)  # accuracy about 10%, on training data, before training!
 
-# ## Model
+#===== TRAINING =====#
 
-# As we mentioned above, a MLP consist of *three* layers that are fully connected. For this example, we define our model with the following layers and dimensions:
+# Make a dataloader using the desired batchsize:
 
-# * **Input:** It has `784` perceptrons (the MNIST image size is `28x28`). We flatten the train and test data so that we can pass them as arguments to this layer.
-# * **Hidden:** It has `32` perceptrons that use the [relu](https://fluxml.ai/Flux.jl/stable/models/nnlib/#NNlib.relu) activation function.
-# * **Output:** It has `10` perceptrons that output the model's prediction or probability that a digit is 0 to 9.
+train_loader = simple_loader(train_data, batchsize = 256)
 
+# Initialise storage needed for the Adam optimiser, with our chosen learning rate:
 
-# We define the model with the `build_model` function:
+opt_state = Flux.setup(Adam(3e-4), model);
 
+# Then train for 30 epochs, printing out details as we go:
 
-function build_model(; imgsize=(28,28,1), nclasses=10)
-    return Chain( Dense(prod(imgsize), 32, relu),
-                  Dense(32, nclasses))
-end
+for epoch in 1:30
+    loss = 0.0
+    for (x, y) in train_loader
+        # Compute the loss and the gradients:
+        l, gs = Flux.withgradient(m -> Flux.crossentropy(m(x), y), model)
+        # Update the model parameters:
+        Flux.update!(opt_state, model, gs[1])
+        # Accumulate the mean loss:
+        loss += l / length(train_loader)
+    end
 
-# Note that we use the functions [Dense](https://fluxml.ai/Flux.jl/stable/models/layers/#Flux.Dense) so that our model is *densely* (or fully) connected and [Chain](https://fluxml.ai/Flux.jl/stable/models/layers/#Flux.Chain) to chain the computation of the three layers.
-
-# ## Loss function
-const loss = logitcrossentropy
-
-# Now, we define the loss function `loss_and_accuracy`. It expects the following arguments:
-# * ADataLoader object.
-# * The `build_model` function we defined above.
-# * A device object (in case we have a GPU available).
-
-function loss_and_accuracy(data_loader, model, device)
-    acc = 0
-    ls = 0.0f0
-    num = 0
-    for (x, y) in data_loader
-        x, y = device(x), device(y)
-        ŷ = model(x)
-        ls += loss(ŷ, y, agg=sum)
-        acc += sum(onecold(ŷ) .== onecold(y)) ## Decode the output of the model
-        num +=  size(x)[end]
+    if mod(epoch, 2) == 1
```

Comment:
```suggestion
    if isodd(epoch)
```
For your consideration.

Reply:
Originally I had a `print_every` variable, and then simplified to hard-code everything. I do think this way points out better how to do something every 10, etc.